File: questions/04_MedianOfTwoSortedArrays/code.js
var findMedianSortedArrays = function(nums1, nums2) {
    const m = nums1.length;
    const n = nums2.length;
    let low = 0, high = m;

    if (m > n) {
        return findMedianSortedArrays(nums2, nums1);
    } //recursive correction reduces code redundancy

    while (low <= high) {
        let partition1 = Math.floor((low + high) / 2);
        let partition2 = Math.floor((m + n + 1) / 2) - partition1;

        let maxLeft1 = (partition1 === 0) ? -Infinity : nums1[partition1 - 1];
        let minRight1 = (partition1 === m) ? Infinity : nums1[partition1];

        let maxLeft2 = (partition2 === 0) ? -Infinity : nums2[partition2 - 1];
        let minRight2 = (partition2 === n) ? Infinity : nums2[partition2];

        if (maxLeft1 <= minRight2 && maxLeft2 <= minRight1) {
            // if we have found the correct partition
            if ((m + n) % 2 === 0) {
                return (Math.max(maxLeft1, maxLeft2) + Math.min(minRight1, minRight2)) / 2;
            } else {
                return Math.max(maxLeft1, maxLeft2);
            }
        } else if (maxLeft1 > minRight2) {
            high = partition1 - 1;
        } else {
            low = partition1 + 1;
        }
    }

    throw new Error("Input arrays are not sorted.");
};

File: questions/04_MedianOfTwoSortedArrays/code.go
package main

import (
    "fmt"
    "math" //MinInt32 and MaxInt32 for edge cases.
)

func findMedianSortedArrays(nums1 []int, nums2 []int) float64 {
	m, n := len(nums1), len(nums2) // m and n defined per question

	
	if m > n { // Ensure nums1 is the smaller array
		return findMedianSortedArrays(nums2, nums1) //recurse
	}

	low, high := 0, m //define low and high for binary search

	for low <= high {
		partitionM := (low + high) / 2
		partitionN := (m + n + 1) / 2 - partitionM

		// Edge cases: handle min/max values when partitioning
		maxLeftM := math.MinInt32
		if partitionM > 0 {
			maxLeftM = nums1[partitionM-1]
		}

		minRightM := math.MaxInt32
		if partitionM < m {
			minRightM = nums1[partitionM]
		}

		maxLeftN := math.MinInt32
		if partitionN > 0 {
			maxLeftN = nums2[partitionN-1]
		}

		minRightN := math.MaxInt32
		if partitionN < n {
			minRightN = nums2[partitionN]
		}

		// Check if correct partition is found
		if maxLeftM <= minRightN && maxLeftN <= minRightM {
			if (m+n)%2 == 0 { // If the total number of elements is even
				return float64(max(maxLeftM, maxLeftN)+min(minRightM, minRightN)) / 2.0
			} else {		
				return float64(max(maxLeftM, maxLeftN)) // If the total number of elements is odd
			}
		} else if maxLeftM > minRightN { //If maxLeftM is too large 
			high = partitionM - 1 //move high left.
		} else { 
			low = partitionM + 1 //move low right
		}
	}

	panic("Input arrays are not sorted or valid") // This should never happen *fingers crossed*
}